pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -237,44 +271,202 @@ func (c *addonDeployController) applyWork(ctx context.Context, required *workapi
+				Manifests: manifests,
 			},
 		},
 	}
+}
 
-	for _, object := range objects {
-		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
-		if err != nil {
-			return nil, err
+func newManifestConfigOption(obj *unstructured.Unstructured) workapiv1.ManifestConfigOption {
+	var resource string
+	var jsonPath workapiv1.JsonPath
+	gvk := obj.GroupVersionKind()
+	switch gvk.Kind {
+	case "Job":
+		resource = "jobs"
+		jsonPath = workapiv1.JsonPath{
+			Name: "isCompleted",
+			Path: `.status.conditions[?(@.type=="Complete")].status`,
 		}
-		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, workapiv1.Manifest{
-			RawExtension: runtime.RawExtension{Raw: rawObject},
-		})
+	case "Pod":
+		resource = "pods"
+		jsonPath = workapiv1.JsonPath{
+			Name: "isReady",
+			Path: `.status.conditions[?(@.type=="Ready")].status`,
+		}
+	default:
+		return workapiv1.ManifestConfigOption{}
+	}
+	return workapiv1.ManifestConfigOption{
+		ResourceIdentifier: workapiv1.ResourceIdentifier{
+			Group:     gvk.Group,
+			Resource:  resource,
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		},
+		FeedbackRules: []workapiv1.FeedbackRule{
+			{
+				Type:      workapiv1.JSONPathsType,
+				JsonPaths: []workapiv1.JsonPath{jsonPath},
+			},
+		},
+	}
+}
+func newHookManifestWork(
+	workName, addonName, clusterName string,
+	manifests []workapiv1.Manifest,
+	manifestConfigs []workapiv1.ManifestConfigOption) *workapiv1.ManifestWork {
+	hookWork := &workapiv1.ManifestWork{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workName,
+			Namespace: clusterName,
+			Labels: map[string]string{
+				constants.AddonLabel: addonName,
+			},
+		},
+		Spec: workapiv1.ManifestWorkSpec{
+			ManifestConfigs: manifestConfigs,
+			Workload: workapiv1.ManifestsTemplate{
+				Manifests: manifests,
+			},
+		},
 	}
 
-	return work, nil
+	return hookWork
 }
 
-func ManifestsEqual(new, old []workapiv1.Manifest) bool {
-	if len(new) != len(old) {
+func isPreDeleteHookObject(object *unstructured.Unstructured) bool {
+	kind := object.GroupVersionKind().Kind
+	if kind != "Job" && kind != "Pod" {
 		return false
 	}
+	labels := object.GetLabels()
+	if _, ok := labels[constants.PreDeleteHookLabel]; ok {
+		return true
+	}
+	return false
+}
 
-	for i := range new {
-		if !equality.Semantic.DeepEqual(new[i].Raw, old[i].Raw) {
+func hookWorkIsFinished(hookWork *workapiv1.ManifestWork) bool {
+	if hookWork == nil {
+		return false
+	}
+	if !meta.IsStatusConditionTrue(hookWork.Status.Conditions, workapiv1.WorkApplied) {
+		return false
+	}
+	if !meta.IsStatusConditionTrue(hookWork.Status.Conditions, workapiv1.WorkAvailable) {
+		return false
+	}
+	for _, manifest := range hookWork.Status.ResourceStatus.Manifests {
+		values := manifest.StatusFeedbacks.Values
+		if len(values) == 0 {
+			return false
+		}
+		if *values[0].Value.String != "True" {
 			return false
 		}
 	}
 	return true
 }
+
+func (c *addonDeployController) runHookWork(
```

Comment:
why not using applyWork here?

Reply:
created another controller to do the pre delete jobs. and use the applyWork to apply hook manifests.